Prevent bots from submitting runs

sivabalansm opened the pull request:
Added Captcha and hidden fields to stop bots from using forms on public instances.
closes #291

sivabalansm committed:
Add spam fields

sivabalansm committed:
Change to about:blank instead of no where lese

sivabalansm committed:
Add captcha gate

sivabalansm committed:
Configure captcha gate to work with my keys

jmlord committed:
moved SpamField to form
+ more consistent naming of id/name
+ htmlFor

jmlord committed:
Removed fallback in favor of finite states
+ added missing useEffect dependencies

jmlord committed:
Moved captcha to form

jmlord committed:
Switching to GEO BON keys.
Public key in code for now, needs to be moved to .env

sivabalansm committed:
Dynamic checking of keys, only verify site captcha if they are available

sivabalansm committed:
Add CSS to hide reCaptcha badge

sivabalansm committed:
Merge remote-tracking branch 'origin/edge' into 291-prevent-bots-from-submitting-runs

jmlord commented on script-server/src/main/kotlin/org/geobon/server/plugins/Routing.kt:
```diff
@@ -292,6 +293,89 @@ fun Application.configureRouting() {
+                if (secretKey.isNullOrEmpty()) {
+                    logger.warn("RECAPTCHA_SERVER_KEY environment variable is not set - captcha verification disabled")
+                    call.respond(
+                        HttpStatusCode.BadRequest,
```
C'est pas très grave, mais je me questionnait sur l'usage de bad request pour quelque chose qui c'est pas permis par une configuration système. J'aurais plutôt tendance à mettre "200 OK" car dans le json tu as success=false. C'est une requête réussie, qui indique qu'on ne peut pas t'authentifier car c'est pas configuré.

(plus haut, le manque de token ou le json mal formé, ça c'est vraiment un "bad request")

jmlord commented on script-server/src/main/kotlin/org/geobon/server/plugins/Routing.kt:
```diff
@@ -292,6 +293,89 @@ fun Application.configureRouting() {
+                    } else {
+                        val errorCodes = responseJSON.optJSONArray("error-codes")
+                        logger.debug("reCAPTCHA verification failed: ${errorCodes?.toString()}")
+                        call.respond(gson.toJson(mapOf("success" to false)))
```
Pourquoi ne pas mettre "reCAPTCHA verification failed" dans "error" ?

jmlord commented on ui/src/components/CaptchaGate.jsx:
```diff
@@ -0,0 +1,185 @@
+          
+          loadRecaptchaScript(config.siteKey);
+        } else {
+          console.warn("Failed to fetch captcha config, assuming captcha is disabled");
```
Peut-être transformer ça en commentaire

jmlord commented on ui/src/components/CaptchaGate.jsx:
```diff
@@ -0,0 +1,185 @@
+      try {
+        // Execute recaptcha v3 
+        const token = await window.grecaptcha.execute(siteKey, {
+          action: "homepage",
```
C'est quoi les valeurs possibles de action? Homepage ne me semble pas trop approprié vu qu'on protège le formulaire de "run".

jmlord commented on script-server/src/main/kotlin/org/geobon/server/plugins/Routing.kt:
```diff
@@ -292,6 +293,89 @@ fun Application.configureRouting() {
         }
 
+        get("/api/captcha-config") {
+            val siteKey = System.getenv("RECAPTCHA_SITE_KEY")
```
Sur la branche dans pipeline-repo c'est RECAPTCHA_CLIENT_KEY
Ça ne fonctionne donc pas quand on le configure tel quel. Il faut choisir un des deux!

jmlord reviewed (changes requested):
Great, it works well.

TODOs:
- Documentation missing! This should be documented in the developer documentation.
- Please create the pipeline-repo PR as well and link to this one.
- The captcha logo is not visible anymore with injection... Not sure if we'll be able to solve that one. Maybe I'll contact google support.

jmlord committed:
just "h" instead of "human"

jmlord committed:
Merge remote-tracking branch 'origin/edge' into 291-prevent-bots-from-submitting-runs

sivabalansm committed:
Code review

sivabalansm committed:
Hiding recaptcha badge when not using CaptchaGate

jmlord commented on ui/src/components/CaptchaGate.jsx:
```diff
@@ -15,6 +15,52 @@ export default function CaptchaGate({ children }) {
+          setCaptchaConfig({ enabled: false });
+          setState(States.VERIFIED);
+  
+          hideRecaptchaBadge();
```
plutôt que de répéter hideRecaptchaBadge partout, il pourrait y avoir un useEffect branché sur captchaConfig.

jmlord reviewed (commented):
Je vais tester et approuver à mon retour des fêtes...

Merci!

sivabalansm committed:
Hide captcha badge bottom right when not on page

jmlord committed:
Merge remote-tracking branch 'origin/edge' into 291-prevent-bots-from-submitting-runs

jmlord committed:
Client instead of site, it's more clear.

jmlord committed:
Remove code / condition duplication

jmlord committed:
Always show badge in form when recaptcha enabled

jmlord committed:
local storage expiry for check result

jmlord committed:
executeRecaptcha does not need to be recreated on every rerender
(it only depends on setters)
+ Same for other stateless functions, which were moved out of component.

jmlord committed:
this is unused
